app/controllers/campaign_pages_controller.rb

```diff
@@ -43,6 +55,42 @@ def create
   end
 
   def show
-    @page = CampaignPage.find params[:id]
+    @campaign_page = CampaignPage.find params[:id]
+    if @campaign_page.active == false
+      redirect_to :campaign_pages, notice: "The page you wanted to view has been deactivated."
+    end  
   end
+
+  def edit
+    @campaign_page = CampaignPage.find params[:id]
+  end
+
+  def update
+    @campaign_page = CampaignPage.find params[:id]
+    @widgets = @campaign_page.campaign_pages_widgets
+
+    permitted_params = CampaignPageParameters.new(params).permit
+    permitted_params[:slug] = permitted_params[:title].parameterize
+    permitted_params[:campaign_pages_widgets_attributes] = []
+    params[:widgets].each do |widget_type_name, widget_data|
+      # widget type id is contained in a field called widget_type:
+      widget_type_id = widget_data.delete :widget_type
+      # This will break if there will be two widgets of the same type for the page. If we enable having two of the same widget type per page,
+      # page display order will need to be considered as well for fetching the correct id of the widget.
+      widget = @widgets.find_by(widget_type_id: widget_type_id)
+      permitted_params[:campaign_pages_widgets_attributes].push({
```

Comment:
Does this run into the same problem with `create` where when you're doing a petition you end up accidentally creating an empty checkbox or textarea? Do we need to clean that out? Is that something we should actually handle on the model?